stac_fastapi/pgstac/core.py

```diff
@@ -39,17 +39,113 @@
 class CoreCrudClient(AsyncBaseCoreClient):
     """Client for core endpoints defined by stac."""
 
-    async def all_collections(self, request: Request, **kwargs) -> Collections:
-        """Read all collections from the database."""
+    async def all_collections(  # noqa: C901
+        self,
+        request: Request,
+        bbox: Optional[BBox] = None,
+        datetime: Optional[DateTimeType] = None,
+        limit: Optional[int] = None,
+        # Extensions
+        query: Optional[str] = None,
+        token: Optional[str] = None,
+        fields: Optional[List[str]] = None,
+        sortby: Optional[str] = None,
+        filter: Optional[str] = None,
+        filter_lang: Optional[str] = None,
+        **kwargs,
+    ) -> Collections:
+        """Cross catalog search (GET).
+
+        Called with `GET /collections`.
+
+        Returns:
+            Collections which match the search criteria, returns all
+            collections by default.
+        """
+        query_params = str(request.query_params)
+
+        # Kludgy fix because using factory does not allow alias for filter-lang
+        if filter_lang is None:
+            match = re.search(r"filter-lang=([a-z0-9-]+)", query_params, re.IGNORECASE)
+            if match:
+                filter_lang = match.group(1)
+
+        # Parse request parameters
+        base_args = {
+            "bbox": bbox,
+            "limit": limit,
+            "token": token,
+            "query": orjson.loads(unquote_plus(query)) if query else query,
+        }
+
+        clean = clean_search_args(
+            base_args=base_args,
+            datetime=datetime,
+            fields=fields,
+            sortby=sortby,
+            filter=filter,
+            filter_lang=filter_lang,
+        )
+
+        # Do the request
+        try:
+            search_request = self.post_request_model(**clean)
```

Comment:
🤔 😬 the issue is that here we don't have to the model used, we should add a `collection_get_request_model` attribute to the client or find a way to avoid this

Reply:
If I understand the problem correctly, we are just trying to dump the args into a json that pgstac will accept, but we are using a model for item search that is defined in `stac_pydantic`. Does this mean we want to add a new attribute to the `CoreCrudClient` that is equivalent to `CollectionSearchExtension().GET`?

Reply:
it's kinda weird because we will pass the Client to the CollectionSearch extension, which kinda result in a circular dependency, but that's how it's done for the `search post model` https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/types/stac_fastapi/types/core.py#L346

> Does this mean we want to add a new attribute to the CoreCrudClient that is equivalent to CollectionSearchExtension().GET?

But yes I think we will have to do that!

Reply:
I tried adding a `collections_get_request_model` attribute based on the `BaseCollectionSearchGetRequest` class but it doesn't work because that class is not a pydantic model. It looks like the `post_request_model` is a `BaseSearchPostRequest` which inherits from `stac_pydantic.api.Search` (which is based on a BaseModel), but `BaseSearchGetRequest` is based on the `APIRequest` class which does not inherit from a pydantic model and thus has no `model_dump_json()` method.

Here is what I tried:
```python

@attr.s
class CoreCrudClient(AsyncBaseCoreClient):
    """Client for core endpoints defined by stac."""

    collections_get_request_model: BaseCollectionSearchGetRequest = attr.ib(
        default=BaseCollectionSearchGetRequest
    )

    async def all_collections(  # noqa: C901
        self,
        request: Request,
        # Extensions
        bbox: Optional[BBox] = None,
        datetime: Optional[DateTimeType] = None,
        limit: Optional[int] = None,
        query: Optional[str] = None,
        token: Optional[str] = None,
        fields: Optional[List[str]] = None,
        sortby: Optional[str] = None,
        filter: Optional[str] = None,
        filter_lang: Optional[str] = None,
        **kwargs,
    ) -> Collections:
        """Cross catalog search (GET).

        Called with `GET /collections`.

        Returns:
            Collections which match the search criteria, returns all
            collections by default.
        """

        # Parse request parameters
        base_args = {
            "bbox": bbox,
            "limit": limit,
            "token": token,
            "query": orjson.loads(unquote_plus(query)) if query else query,
        }

        clean = clean_search_args(
            base_args=base_args,
            datetime=datetime,
            fields=fields,
            sortby=sortby,
            filter=filter,
            filter_lang=filter_lang,
        )

        # Do the request
        try:
            search_request = self.collections_get_request_model(**clean)
        except ValidationError as e:
            raise HTTPException(
                status_code=400, detail=f"Invalid parameters provided {e}"
            ) from e

        return await self._collection_search_base(search_request, request=request)


```

So I think I will need to get the json for the pgstac query using a different approach.

Reply:
I got this working by using `BaseCollectionSearchPostRequest` (instead of `BaseCollectionSearchGetRequest`) which follows the pattern used in the item search.

Reply:
for get request I think we should use the GET model or it might not work properly! 

Maybe I was wrong before and looking at the code, I think we can use `PgstacSearch` at line 91 

or we could create another pydantic model which match what `pgstac.collection_search` expect

Reply:
Ah, I think I see the path. I probably need to use `create_request_model` to make sure the appropriate args get added for the extensions. However, without creating a new pydantic model, I would still need to use the `POST` model because the search functionality relies on the pydantic model. That's probably something that belongs in stac-fastapi, right? Is there a particular reason the `GET` request models are a generic class rather than a pydantic model in stac-fastapi?

Reply:
>  Is there a particular reason the GET request models are a generic class rather than a pydantic model in stac-fastapi?

yes because pydantic model are interpreted as body parameter for endpoints

Reply:
Thanks for your patience while I figure this out. I think I need to double back and make sure I understand how all of these models are getting used. Upon closer examination, it looks like stac-fastapi-pgstac is just using the `AsyncBaseCoreClient.post_request_model` for all of the POST requests:
e.g.
https://github.com/stac-utils/stac-fastapi-pgstac/blob/cb08b0f6f1b11643632071113e10921501da3808/stac_fastapi/pgstac/core.py#L324

We provide a `search_post_request_model` to `StacAPI`:
https://github.com/stac-utils/stac-fastapi-pgstac/blob/cb08b0f6f1b11643632071113e10921501da3808/stac_fastapi/pgstac/app.py#L67-L78

But I can't see how `search_post_request_model` is getting used by anything right now.

Reply:
I guess `search_get_request_model` and `search_post_request_model` are a dead end. We pass the `post_request_model` that is based on `PgstacSearch` to the `CoreCrudClient` and that is what gets used to generate inputs for all of the `pgstac` search functions. 

I am pretty sure we can use it in the same way for the `/collections` GET method like we do in the `/collections/{collection_id}/items` GET method.